cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,84 @@ class DirichletBC
-          for (std::size_t i = 0; i < _dofs0.size(); ++i)
-          {
-            if (_dofs0[i] < x_size)
+        assert(x.size() <= x0->size());
+        apply(
+            [dofs_g, x0 = *x0, alpha, values,
+             &dofs0 = this->_dofs0](std::int32_t i) -> T
             {
-              assert(dofs1_g[i] < (std::int32_t)values.size());
-              x[_dofs0[i]] = alpha * (values[dofs1_g[i]] - _x0[_dofs0[i]]);
-            }
-          }
-        }
-        else
-        {
-          for (std::size_t i = 0; i < _dofs0.size(); ++i)
-          {
-            if (_dofs0[i] < x_size)
+              assert(dofs_g[i] < static_cast<std::int32_t>(values.size()));
+              return alpha * (values[dofs_g[i]] - x0[dofs0[i]]);
+            });
+      }
+      else
+      {
+        apply(
+            [dofs_g, values, alpha](std::int32_t i) -> T
             {
-              assert(dofs1_g[i] < (std::int32_t)values.size());
-              x[_dofs0[i]] = alpha * values[dofs1_g[i]];
-            }
-          }
-        }
+              assert(dofs_g[i] < static_cast<std::int32_t>(values.size()));
+              return alpha * values[dofs_g[i]];
+            });
       }
-      else if (std::holds_alternative<std::shared_ptr<const Constant<T>>>(_g))
+    };
+
+    // called by std::visist in case g holds a Constant
+    auto handle_constant = [&](std::shared_ptr<const Constant<T>> g)
+    {
+      const std::vector<T>& value = g->value;
+      std::int32_t bs = _function_space->dofmap()->bs();
+      if (x0)
       {
-        auto g = std::get<std::shared_ptr<const Constant<T>>>(_g);
-        const std::vector<T>& value = g->value;
-        std::int32_t bs = _function_space->dofmap()->bs();
-        if (x0.has_value())
-        {
-          assert(x.size() <= x0.value().size());
-          std::ranges::for_each(
-              _dofs0,
-              [x_size, &x, x0 = x0.value(), &value, alpha, bs](auto dof)
-              {
-                if (dof < x_size)
-                  x[dof] = alpha * (value[dof % bs] - x0[dof]);
-              });
-        }
-        else
-        {
-          std::ranges::for_each(_dofs0,
-                                [x_size, bs, alpha, &value, &x](auto dof)
-                                {
-                                  if (dof < x_size)
-                                    x[dof] = alpha * value[dof % bs];
-                                });
-        }
+        assert(x.size() <= x0->size());
+        apply(
+            [x0 = *x0, alpha, bs, &value, &dofs0 = _dofs0](std::int32_t i) -> T
+            {
+              auto dof = dofs0[i];
+              return alpha * (value[dof % bs] - x0[dof]);
+            });
       }
-    }
+      else
+      {
+        apply([alpha, bs, &value, &dofs0 = _dofs0](std::int32_t i) -> T
+              { return alpha * value[dofs0[i] % bs]; });
+      }
+    };
+
+    std::visit(overloaded{handle_function, handle_constant}, _g);
```

Comment:
I think using `std::visit` brings in more obfuscation than is necessary. We only have two cases, so an if-else switch would be clearer, would allow control of the types passed to the lambda functions, and would avoid the 'overloaded' boiler plate code further up in the file.

Reply:
Unfortunate, but I get the point and it is really the only occasion where `std::variant` would be beneficial within dolfinx. Will remove.